azbus/disposition.go

```diff
@@ -51,77 +56,139 @@ func (r *Receiver) dispose(ctx context.Context, d Disposition, err error, msg *R
 	defer span.Finish()
 	log.Infof("DeadLetter Message: %v", err)
 	options := azservicebus.DeadLetterOptions{
 		Reason: to.Ptr(err.Error()),
 	}
-	err1 := r.receiver.DeadLetterMessage(ctx, msg, &options)
+	err1 := r.DeadLetterMessage(ctx, msg, &options)
 	if err1 != nil {
 		azerr := fmt.Errorf("DeadLetter Message failure: %w", NewAzbusError(err1))
 		log.Infof("%s", azerr)
 	}
 }
 
-func (r *Receiver) complete(ctx context.Context, err error, msg *ReceivedMessage) {
+func complete(ctx context.Context, log logger.Logger, r *azservicebus.Receiver, err error, msg *ReceivedMessage) {
 	ctx = context.WithoutCancel(ctx)
-	log := r.log.FromContext(ctx)
-	defer log.Close()
 
 	span, _ := tracing.StartSpanFromContext(ctx, "Message.Complete")
 	defer span.Finish()
 
 	if err != nil {
 		log.Infof("Complete Message %v", err)
 	} else {
-		log.Infof("Complete Message")
+		log.Debugf("Complete Message")
 	}
 
-	err1 := r.receiver.CompleteMessage(ctx, msg, nil)
+	err1 := r.CompleteMessage(ctx, msg, nil)
 	if err1 != nil {
 		// If the completion fails then the message will get rescheduled, but it's effect will
 		// have been made, so we could get duplication issues.
 		azerr := fmt.Errorf("Complete: failed to settle message: %w", NewAzbusError(err1))
 		log.Infof("%s", azerr)
 	}
 }
+
+// Reschedule handles when a message should be deferred at a later time. There are a
+// number of ways of doing this but it turns out that simply not doing anything causes
+// azservicebus to resubmit the message 1 minute later. We keep the function signature with
+// unused arguments for consistency and in case we need to implement more sophisticated
+// algorithms in future.
+func reschedule(ctx context.Context, log logger.Logger, r *azservicebus.Receiver, err error, msg *ReceivedMessage) {
+	ctx = context.WithoutCancel(ctx)
+
+	span, _ := tracing.StartSpanFromContext(ctx, "Message.Reschedule")
+	defer span.Finish()
+	log.Infof("Reschedule Message on DeliveryCount %d: %v", msg.DeliveryCount, err)
+}
+
+// Abandon abandons message. This function is not used but is present for consistency.
+func (r *Receiver) abandon(ctx context.Context, err error, msg *ReceivedMessage) {
+	log := r.log.FromContext(ctx)
+	defer log.Close()
+
+	abandon(ctx, log, r.receiver, err, msg)
+}
+
+func (r *Receiver) reschedule(ctx context.Context, err error, msg *ReceivedMessage) {
+	log := r.log.FromContext(ctx)
+	defer log.Close()
+
+	reschedule(ctx, log, r.receiver, err, msg)
+}
+
+func (r *Receiver) deadLetter(ctx context.Context, err error, msg *ReceivedMessage) {
+	log := r.log.FromContext(ctx)
+	defer log.Close()
+	deadLetter(ctx, log, r.receiver, err, msg)
+}
+
+func (r *Receiver) complete(ctx context.Context, err error, msg *ReceivedMessage) {
+	log := r.log.FromContext(ctx)
+	defer log.Close()
+
+	complete(ctx, log, r.receiver, err, msg)
+}
+
+// Abandon abandons message. This function is not used but is present for consistency.
```

Comment:
i could split this file out as well, but there seems to be some value in having all the disposition stuff in one place